File: DataStructures/Java/DoublyLinkedList.java
public class DoublyLinkedList {
    private Node head;
    private Node tail;

    static class Node {
        // CSOFF: VisibilityModifier
        int data;
        Node next;
        Node prev;
        // CSON: VisibilityModifier

        Node(final int newData) {
            this.data = newData;
            this.next = null;
            this.prev = null;
        }
    }

    public DoublyLinkedList() {
        this.head = null;
        this.tail = null;
    }

    public final void insertAtLast(final int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            return;
        }
        this.tail.next = newNode;
        newNode.prev = this.tail;
        this.tail = newNode;
    }

    public final void insertAtFirst(final int data) {
        Node newNode = new Node(data);
        newNode.next = this.head;
        this.head.prev = newNode;
        this.head = newNode;
    }

    public final void insertAfter(final int data, final int after) {
        Node newNode = new Node(data);
        Node temp = this.head;
        while (temp != null) {
            if (temp.data == after) {
                newNode.next = temp.next;
                newNode.prev = temp;
                temp.next = newNode;
                return;
            }
            temp = temp.next;
        }
    }

    public final void print() {
        Node temp = this.head;
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public final void delete(final int data) {
        if (this.head == null) {
            return;
        }
        if (this.head.data == data) {
            this.head = this.head.next;
            this.head.prev = null;
            return;
        }
        Node temp = this.head;
        while (temp.next != null) {
            if (temp.next.data == data) {
                temp.next = temp.next.next;
                if (temp.next != null) {
                    temp.next.prev = temp;
                }
                return;
            }
            temp = temp.next;
        }
    }

    public static void main(final String[] args) {
        DoublyLinkedList dll = new DoublyLinkedList();
        for (int i : new int[] { 1, 2, 3, 4, 5 }) {
            dll.insertAtLast(i);
        }
        dll.print();
    }
}

File: DataStructures/C/doubly_linked_list.c
// Doubly Linked List

#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int value;
    struct Node *next;
};

struct Node *head = NULL;
struct Node *tail = NULL;

void insert_at_last(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    if (head == NULL)
    {
        new_node->value = data;
        new_node->next = new_node;
        head = new_node;
        tail = new_node;
    }
    else
    {
        new_node->value = data;
        new_node->next = head;
        tail->next = new_node;
        tail = new_node;
    }
}

void insert_at_first(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    new_node->value = data;
    new_node->next = head;
    head = new_node;
    tail->next = head;
}

void delete_node(int data)
{
    if (head == NULL)
    {
        printf("List is empty\n");
        return;
    }

    struct Node *temp = head;
    struct Node *prev = NULL;

    if (head->value == data)
    {
        head = head->next;
        tail->next = head;
        free(temp);
        return;
    }

    while (temp->value != data)
    {
        if (temp->next == head)
        {
            printf("Element not found\n");
            return;
        }
        prev = temp;
        temp = temp->next;
    }

    prev->next = temp->next;
    free(temp);
}

void display()
{
    if (head == NULL)
    {
        printf("List is empty\n");
        return;
    }

    struct Node *temp = head;
    do
    {
        printf("%d ", temp->value);
        temp = temp->next;
    } while (temp != head);
    printf("\n");
}

int main()
{
    insert_at_last(1);
    insert_at_last(2);
    insert_at_last(3);
    insert_at_last(4);
    insert_at_last(5);
    insert_at_first(0);
    delete_node(3);
    display();
    return 0;
}